### PR TITLE
Feat / injectable components

### PR DESCRIPTION
**Pull request by royschut**
## Injectable components
This PR adds functionality to replace ui components in `ui-react` externally. This is useful to replace components for specific platforms (such as the player in capacitor) or if clients want to easily replace components. [OTT-236]

- I've used Inversify, to align with other overrides we're doing already. However, Inversify is built for classes and const values, so I had to be a bit creative in order to make it work for functional react components (without adding a new dep such as `inversify-react`). 
- I decided to use a wrapper function, so that the component itself checks if an override is set. Initially I did this from where the component is used, or even higher up in the react hierarchy. However, I'm a bit scared that we might accidentally import the component somewhere, accidentally ignoring the configured overrides. So I found this approach the cleanest and safest.
- Initially I also bound the default component from the ui-react container, but I found this becoming a bit bloated. Curious what you think about this approach.
- Perhaps more of my approach is opinionated. Feel free to share your thoughts!

[OTT-236]: https://videodock.atlassian.net/browse/OTT-236?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/ui-react/src/components/Card/Card.tsx [hunk 1]
```diff
@@ -43,6 +44,8 @@ function Card({
   url,
 }: CardProps): JSX.Element {
   const { title, duration, episodeNumber, seasonNumber, cardImage: image, mediaStatus, scheduledStart } = item;
+  // const Tag = container.get<React.FC<TagProps>>(TAG_COMPONENT);
```
Comment (royschut) [inline packages/ui-react/src/components/Card/Card.tsx]: This is obsolete code, but I will remove this whole example commit later.

File: packages/ui-react/src/modules/InjectableComponent.tsx
Comment (royschut) [inline packages/ui-react/src/modules/InjectableComponent.tsx:1]: I'm not sure if this is the correct folder for this. (I'm even not sure if a `modules` folder is correct for this container)

File: packages/ui-react/src/modules/container.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import 'reflect-metadata'; // include once in the app for inversify (see: https://github.com/inversify/InversifyJS/blob/master/README.md#-installation)
```
Comment (royschut) [inline packages/ui-react/src/modules/container.ts]: Perhaps this is not necessary, given that it's done from `common` as well?

**Comment by langemike** [comment]
> Initially I also bound the default component from the ui-react container, but I found this becoming a bit bloated. Curious what you think about this approach.

I like the current approach. And I do not see any improvements we could add. So I would say, really nice job and I approve!

**Review by langemike — APPROVED** [review]

**Review by ChristiaanScheermeijer — COMMENTED** [review]
Nice @royschut! I have some feedback and suggestions, mainly to prevent spreading the props and types in different places. But this is opinionated, of course.

Your change introduces a new pattern for overriding components via an `InjectableComponent` (which is a component). The only reason for this is to prevent a HoC, but we are now essentially creating a HoC in each component that we want to override. I suggest that we make it a HoC, so we can clean up the redundant function wrapper in each component:

**InjectableComponent.tsx**
```tsx
export const createInjectableComponent = <Props extends object>(identifier: symbol, DefaultComponent: React.FC<Props>) => {
  return (props: Props) => {
    const isOverridden = container.isBound(identifier);

    if (isOverridden) {
      const OverriddenComponent = container.get<React.FC<Props>>(identifier);

      return <OverriddenComponent {...props} />;
    }

    return <DefaultComponent {...props} />;
  };
};
```

**Tag.tsx**
```tsx
export default createInjectableComponent(identifier, Tag);
```

I'm no fan of moving the component prop typings to the types folder. Why don't we export the prop typings from the component file directly?

Last thing, we should ensure that the overridden components match the expected type. We can do this manually like so:

```ts
uiComponentContainer.bind<React.FC<TagProps>>(tagIdentifier).toConstantValue(Tag);
```

But a better solution is not to work with `container` directly. For example, we can create helper functions where we require the type (or infer it).

**Comment by royschut** [comment]
> I have some feedback and suggestions

Thanks a lot @ChristiaanScheermeijer!

I agree with all 3 your suggestions. I will improve it. 
(You mentioned preventing spreads, but the amount of spreads is the same. :) However I prefer to use a standard pattern indeed: HOC.)

File: packages/ui-react/src/modules/container.ts (hunk 2 above)
Comment (R-Cloud-Designs) [reply packages/ui-react/src/modules/container.ts]: Considering it's a singleton and we import the package from `common`, I think we can skip this. The Inversify docs caution against importing the package multiple times, as `Reflect` should be a global singleton.

Comment (royschut) [reply packages/ui-react/src/modules/container.ts]: Thanks Rachid!

**Review by ChristiaanScheermeijer — APPROVED** [review]
Nice! We might need some more changes, but let's give it a spin first 😄

**Review by MelissaDTH — APPROVED** [review]

**Pull request merged**